contracts/eosio.system/src/finalizer_key.cpp

```diff
@@ -48,51 +48,74 @@ namespace eosiosystem {
             continue;
          }
 
-         // stores finalizer_key_id
-         first_finalizer_key_set.emplace_back(finalizer->active_finalizer_key_id);
-
-         // builds up finalizer_authorities
-         finalizer_authorities.emplace_back(
-            eosio::finalizer_authority{
+         proposed_fin_keys.emplace_back( proposed_finalizer_key_t {
+            .key_id        = finalizer->active_finalizer_key_id,
+            .fin_authority = eosio::finalizer_authority{
                .description = finalizer->finalizer_name.to_string(),
                .weight      = 1,
                .public_key  = finalizer->active_finalizer_key_binary
-            }
+            }}
          );
       }
 
-      check( finalizer_authorities.size() > 0 && finalizer_authorities.size() >= _gstate.last_producer_schedule_size, "not enough top producers have registered finalizer keys, has " + std::to_string(finalizer_authorities.size()) );
+      check( proposed_fin_keys.size() == _gstate.last_producer_schedule_size,
+            "not enough top producers have registered finalizer keys, has " + std::to_string(proposed_fin_keys.size()) + ", require " + std::to_string(_gstate.last_producer_schedule_size) );
 
-      set_finalizers(std::move(finalizer_authorities), first_finalizer_key_set, {});
+      set_proposed_finalizers_sorted(proposed_fin_keys);
       check( is_savanna_consensus(), "switching to Savanna failed" );
    }
 
-   // This function may never fail, as it can be called by update_elected_producers,
-   // and in turn by onblock
-   void system_contract::set_finalizers( std::vector<eosio::finalizer_authority>&& finalizer_authorities, const std::vector<uint64_t>& new_ids, const std::unordered_set<uint64_t>& kept_ids ) {
+   // This function may never fail, as it can be called by onblock indirectly.
+   // Establish finalizer policy from `proposed_fin_keys` and calls
+   // eosio::set_finalizers host function
+   void system_contract::set_proposed_finalizers( const std::vector<proposed_finalizer_key_t>& proposed_fin_keys ) {
```

Comment:
I think this should always do the sort by `key_id` and a check to ensure things have changed as `set_proposed_finalizers_if_changed` does. It should also always write it out to the table if it is going to call the `set_finalizers` host function, against as `set_proposed_finalizers_if_changed` does. 

That would mean you could then remove `set_proposed_finalizers_if_changed` and `set_proposed_finalizers_sorted`.

Yes, this means that in some situations there is a redundant sort and a redundant (in)equality check. But that is okay. That is relatively low computation cost (and only unnecessarily paid for in the `switchtosvnn` and some `actfinkey` actions), but it gives us more confidence that the correct thing is going to happen regardless of what the caller of the function does.

Reply:
Thanks. This will be safer.